### PR TITLE
Sim and Hash functions for Boolean Datatype

### PR DESCRIPTION
**Pull request by sania-16**
have added similarity and hash functions issue #921 
please review @sonalgoyal 
need to work on integrating and testing them

### CODE REVIEW

File: common/core/src/main/java/zingg/common/core/hash/HashFnBoolean.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package zingg.common.core.hash;
+
+public class HashFnBoolean extends BaseHash<Boolean,Integer>{
+
+	public HashFnBoolean() {
+	    setName("HashFnBoolean");
+	}
+
+    public Integer call(Boolean field){
+        return Boolean.hashCode(field); // 1231 for true and 1237 for false
```
Comment (sonalgoyal) [inline common/core/src/main/java/zingg/common/core/hash/HashFnBoolean.java:10]: no need for this hash fucntion

File: common/core/src/main/java/zingg/common/core/similarity/function/BooleanSimilarityFunction.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+	public static final Log LOG = LogFactory
+			.getLog(BooleanSimilarityFunction.class);
+
+	public BooleanSimilarityFunction() {
+		super("BooleanSimilarityFunction");
+	}
+    
+    @Override
+    public Double call(Boolean first, Boolean second) {
+		if (first == null && second == null) return 1d;
```
[the remark this answers is not in the document]
Comment (sonalgoyal) [reply common/core/src/main/java/zingg/common/core/similarity/function/BooleanSimilarityFunction.java:18]: there is a common generic class similarity function which has all this logic defined. you can use it directly.

File: common/core/src/main/java/zingg/common/core/hash/HashFnBoolean.java (hunk 1 above)
Comment (sania-16) [reply common/core/src/main/java/zingg/common/core/hash/HashFnBoolean.java:10]: will remove this class completely

File: common/core/src/main/java/zingg/common/core/similarity/function/BooleanSimilarityFunction.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (sania-16) [reply common/core/src/main/java/zingg/common/core/similarity/function/BooleanSimilarityFunction.java:18]: will integrate it inside the call function

**Comment by sania-16** [comment]
sim and hash functions need to be added as well to the registries - hashfunctionregistry and featurefactory

File: common/core/src/main/java/zingg/common/core/feature/BooleanFeature.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+public class BooleanFeature extends BaseFeature<Boolean> {
+
+    private static final long serialVersionUID = 1L;
+
+	public BooleanFeature() {
+	}
+
+    public void init(FieldDefinition f){
+        setFieldDefinition(f);
+        if (f.getMatchType().contains(MatchType.FUZZY)) {
```
Comment (sonalgoyal) [inline common/core/src/main/java/zingg/common/core/feature/BooleanFeature.java:18]: dont think there is a use for fuzzy in boolean. please remove

File: common/core/src/main/java/zingg/common/core/similarity/function/BooleanSimilarityFunction.java [hunk 4]
```diff
@@ -0,0 +1,20 @@
+package zingg.common.core.similarity.function;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+public class BooleanSimilarityFunction extends SimilarityFunctionExact<Boolean>{
+
+    private static final long serialVersionUID = 1L;
+	public static final Log LOG = LogFactory.getLog(BooleanSimilarityFunction.class);
```
[the remark this answers is not in the document]
Comment (sonalgoyal) [reply common/core/src/main/java/zingg/common/core/similarity/function/BooleanSimilarityFunction.java:9]: you dont need a class for this type. see how we are doing it for dates in the DateFeature. addSimFunction(new SimilarityFunctionExact<Date>("DateSimilarityFunctionExact"));

**Review by sonalgoyal — CHANGES_REQUESTED** [review]
see comments.

also add BooleanFeature to the SparkFeatureFactory

File: common/core/src/main/java/zingg/common/core/similarity/function/BooleanSimilarityFunction.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (sania-16) [reply common/core/src/main/java/zingg/common/core/similarity/function/BooleanSimilarityFunction.java:9]: @sonalgoyal 
For DateFeature, we do have DateSimilarityFunction.java defined
So do we need this class?

**Pull request merged**